### PR TITLE
Docs for prompt engineering UI

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Docs for prompt engineering UI

## How is this patch tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [X] Instructions

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Comment by BenWilson2** [comment]
@mlflow-automation autoformat

File: docs/source/llms/prompt-engineering.rst [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+The prompt engineering UI also requires a connection between the AI Gateway and the MLflow
+Tracking Server. To connect the AI Gateway with the MLflow Tracking Server, simply set the
+``MLFLOW_AI_GATEWAY`` environment variable in the environment where the server is running and
+restart the server. For example, if the AI Gateway is running at ``http://localhost:7000``, you
+can start an MLflow Tracking Server in a shell on your local machine and connect it to the
+AI Gateway using the :ref:`mlflow server <cli>` command as follows:
+
+.. code-block:: bash
+
+   export MLFLOW_AI_GATEWAY="http://localhost:7000"
```
Comment (harupy) [inline docs/source/llms/prompt-engineering.rst:42]:
```suggestion
   export MLFLOW_GATEWAY_URI="http://localhost:7000"
```

Comment (harupy) [reply docs/source/llms/prompt-engineering.rst:42]:
```suggestion
   mlflow gateway start --config-path path/to/config.yaml --port 7000
   export MLFLOW_AI_GATEWAY="http://localhost:7000"
```

I'd also add `mlflow gateway start` before `MLFLOW_GATEWAY_URI`.

File: docs/source/llms/prompt-engineering.rst [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+      {
+          "dataframe_records": [
+              {
+                  "article": "An MLflow Project is a format for packaging data science code...",
+                  "question": "What is an MLflow Project?"
+              }
+          ]
+      }'
+
+      echo $sample_input | curl \
```
Comment (harupy) [inline docs/source/llms/prompt-engineering.rst:333]:
```suggestion
      echo $input | curl \
```

File: docs/source/llms/prompt-engineering.rst [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+*using Prompt Engineering*. This will open the prompt engineering playground where you can try
+out different LLMs, parameters, and prompts.
+
+|new_run| |prompt_modal_1|
+
+.. |new_run| image:: ../_static/images/new_run.png
+   :width: 25%
+
+.. |prompt_modal_1| image:: ../_static/images/prompt_modal_1.png
+   :width: 70%
```
Comment (harupy) [inline docs/source/llms/prompt-engineering.rst:65]: This part looks like this:

image

File: docs/source/llms/prompt-engineering.rst [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+   unrelated to MLflow to see how the LLM responds. This is important to ensure that the application
+   is robust to irrelevant inputs.
+
+    |add_row| |add_row_modal|
+
+    .. |add_row| image:: ../_static/images/add_row.png
+       :width: 10%
+
+    .. |add_row_modal| image:: ../_static/images/add_row_modal.png
+       :width: 50%
```
Comment (harupy) [inline docs/source/llms/prompt-engineering.rst:154]: This part looks like this. Is this intended?

image

File: docs/source/llms/prompt-engineering.rst [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+for generating product advertisements.
+
+MLflow will embed the specified *stock_type* input
+variable value - ``"books"`` - into the specfied *prompt  template* and send it to the LLM
+associated with the AI Gateway route with the configured *temperature* (currently ``0.01``)
+and *max_tokens* (currently empty). The LLM response will appear in the *Output* section.
+
+.. figure:: ../_static/images/prompt_modal_2.png
+   :scale: 25%
+   :align: center
```
Comment (prithvikannan) [inline docs/source/llms/prompt-engineering.rst:80]: this image has max_tokens of 1000. can we update the text?

**Review by prithvikannan — APPROVED** [review]
LGTM overall, couple small things

File: docs/source/llms/prompt-engineering.rst [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+Quickstart
+==========
+
+The following guide will get you started with MLflow's UI for prompt engineering.
+
+Step 1: Create an AI Gateway Completions Route
+----------------------------------------------
+To use the prompt engineering UI, you need to create one or more :ref:`AI Gateway <gateway>`
+completions :ref:`Routes <routes>`. Follow the
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst]: Great catch - fixed!

File: docs/source/llms/prompt-engineering.rst (hunk 1 above)
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst]: Oof, great catch. Fixed the env var and added a code block for gateway start in Step 1

File: docs/source/llms/prompt-engineering.rst (hunk 3 above)
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst:70]: Yes, let me know if you have suggestions for improvements

File: docs/source/llms/prompt-engineering.rst (hunk 5 above)
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst]: Great catch - fixed!

File: docs/source/llms/prompt-engineering.rst [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+
+*Read the following article from the MLflow documentation that appears between triple
+backticks. Then, answer the question about the documentation that appears between triple quotes.
+Include relevant links and code examples in your answer.*
+
+*```{{article}}```*
+
+*"""*
+*{{question}}*
+*"""*
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst]: Great idea, done!

File: docs/source/llms/prompt-engineering.rst (hunk 4 above)
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst:161]: Yes, let me know if you have suggestions for improvements

File: docs/source/llms/prompt-engineering.rst (hunk 2 above)
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst]: Great catch!

**Review by dbczumar — COMMENTED** [review]
@prithvikannan @harupy thanks for the excellent feedback. Addressed your comments! :)

File: docs/source/llms/prompt-engineering.rst (hunk 3 above)
Comment (harupy) [reply docs/source/llms/prompt-engineering.rst:70]: maybe GIF?

Comment (harupy) [reply docs/source/llms/prompt-engineering.rst:70]: or a screenshot like this:

image

File: docs/source/llms/prompt-engineering.rst [hunk 8]
```diff
@@ -0,0 +1,351 @@
+.. _prompt-engineering:
+
+====================================
+Prompt Engineering UI (experimental)
```
Comment (harupy) [inline docs/source/llms/prompt-engineering.rst:4]: image

```suggestion
Prompt Engineering UI (Experimental)
```

to be consistent

File: docs/source/llms/prompt-engineering.rst [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+
+1. To improve performance, create a new Run by selecting the *Duplicate run* option from the context
+   menu. For example, in the MLflow documentation use case, adding the following text to
+   the prompt template helps improve robustness to irrelevant questions:
+
+   .. code-block::
+
+      If the question does not relate to the article, respond exactly with the phrase
+      "I do not know how to answer that question." Do not include any additional text in your
+      response.
```
Comment (harupy) [inline docs/source/llms/prompt-engineering.rst:191]: image

```suggestion
   .. code-block:: text

      If the question does not relate to the article, respond exactly with the phrase
      "I do not know how to answer that question." Do not include any additional text in your
      response.
```

to disable syntax highlighting on non-python code

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

File: docs/source/llms/prompt-engineering.rst (hunk 9 above)
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst]: Thank you! Fixed!

File: docs/source/llms/prompt-engineering.rst (hunk 8 above)
Comment (dbczumar) [reply docs/source/llms/prompt-engineering.rst]: Thank you! Fixed!

**Pull request merged**